### PR TITLE
Respect tool upgrades in `uv tool install`

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

For now the semantics are such that if the requested requirements from the command line don't match the receipt (or if any `--reinstall` or `--upgrade` is requested), we proceed with an install, passing the `--reinstall` and `--upgrade` to the underlying Python environment.

This may lead to some unintuitive behaviors, but it's simplest for now. For example:

- `uv tool install black<24` followed by `uv tool install black --upgrade` will install the latest version of `black`, removing the `<24` constraint.
- `uv tool install black --with black-plugin` followed by `uv tool install black` will remove `black-plugin`.

Closes https://github.com/astral-sh/uv/issues/4659.

### CODE REVIEW

File: crates/uv/tests/tool_install.rs
Comment (charliermarsh) [inline crates/uv/tests/tool_install.rs:1241]: This is a behavior change but I personally think it's ok to copy over the entrypoints even on a no-op. I could probably figure out a way to avoid this as an optimization though.

File: crates/uv/src/commands/tool/install.rs
Comment (charliermarsh) [inline crates/uv/src/commands/tool/install.rs:181]: It seems ok to me to only remove this when `--force` is provided?

Comment (charliermarsh) [reply crates/uv/src/commands/tool/install.rs:181]: In other words: can't we defer to our underlying virtualenv update machinery to manage this by reinstalling packages as required?

**Comment by charliermarsh** [comment]
Don't understand the dreaded Windows failures yet but I'll look at them once approved.

File: crates/uv/tests/tool_install.rs
Comment (zanieb) [reply crates/uv/tests/tool_install.rs:1241]: I agree it's "okay" but it feels a little weird to me.

Comment (charliermarsh) [reply crates/uv/tests/tool_install.rs:1241]: I think it's safer and simpler to copy them over every time, avoids any weird issues around trying to detect if the package changed in the environment or the entrypoints getting out-of-sync with the environment.

File: crates/uv/src/commands/tool/install.rs
Comment (zanieb) [reply crates/uv/src/commands/tool/install.rs:181]: That seems okay, yeah.

File: crates/uv/src/commands/tool/install.rs
Comment (zanieb) [inline crates/uv/src/commands/tool/install.rs:177]: I don't think `force` needs to imply deletion of the environment though? This seems a little weird. I think we should only remove the existing environment if the requested Python interpreter is different.

Comment (charliermarsh) [reply crates/uv/src/commands/tool/install.rs:177]: We can do that… I feel like force is sorta useful too personally.

Comment (charliermarsh) [reply crates/uv/src/commands/tool/install.rs:177]: Force would be like: completely ignore whatever already exists.

Comment (zanieb) [reply crates/uv/src/commands/tool/install.rs:177]: Hm. It feels weird to mix the purpose for me. I agree removing the environment _sounds_ useful but in what case would someone actually need to do that?

**Review by zanieb — APPROVED** [review]

File: crates/uv/src/commands/tool/install.rs
Comment (zanieb) [reply crates/uv/src/commands/tool/install.rs:177]: (it seems safe to discuss and address this separately from this pull request)

Comment (charliermarsh) [reply crates/uv/src/commands/tool/install.rs]: I guess if the environment was broken and the receipt was out-of-sync, even `uninstall` wouldn't remove it...? Maybe I'll change `uninstall` to handle that edge case, then remove this.

Comment (zanieb) [reply crates/uv/src/commands/tool/install.rs]: If the environment is broken then `--force` totally could / should remove it.

Comment (charliermarsh) [reply crates/uv/src/commands/tool/install.rs]: Maybe: if there's no receipt, and the environment exists, we remove it? :)

**Pull request merged**
